monads/src/FailCont/Main.hs

```diff
@@ -7,13 +7,49 @@ data Error
   = EmptyInput 
   | ParseFailed String 
   | DivisionByZero
+  | NotPrime Int
   deriving (Show, Eq)
 
 addInts :: String -> String -> FailCont r Error Int
-addInts = undefined 
+addInts "" _ = FailCont $ \f -> f (Left EmptyInput)
+addInts _ "" = FailCont $ \f -> f (Left EmptyInput)
+addInts s1 s2 = FailCont $ \f ->
+  case readMaybe s1 of
+    Nothing -> f (Left (ParseFailed s1))
+    Just x -> case readMaybe s2 of
+      Nothing -> f (Left (ParseFailed s2))
+      Just y -> f (Right (x + y))
```

Comment:
Create a helper function that does `readMaybe` and returns `Either`. This way, your code will be way more beautiful.